// @sveltepress/theme-default/dist/components/layout.js
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { get, writable } from 'svelte/store';
import themeOptions from 'virtual:sveltepress/theme-default';
export var MOBILE_EDGE_WIDTH = 950;
export var sidebarCollapsed = writable(true);
export var tocCollapsed = writable(true);
export var navCollapsed = writable(true);
export var anchors = writable([]);
export var pages = writable([]);
export var scrollY = writable(0);
export var oldScrollY = writable(0);
export var scrollDirection = writable('up');
export var isDark = writable(false);
export var sidebar = writable(true);
export var showHeader = writable(true);
export var resolvedSidebar = writable(Object.entries((themeOptions.sidebar || {})).reduce(function (all, _a) {
    var item = _a[1];
    return __spreadArray(__spreadArray([], all, true), item, true);
}, []));
scrollY.subscribe(function (sy) {
    var nextDirection = sy - get(oldScrollY) > 0 ? 'down' : 'up';
    if (nextDirection !== get(scrollDirection))
        scrollDirection.set(nextDirection);
});
resolvedSidebar.subscribe(function (sidebar) {
    pages.set(sidebar.reduce(function (allPages, item) {
        return Array.isArray(item.items)
            ? __spreadArray(__spreadArray([], allPages, true), item.items, true) : __spreadArray(__spreadArray([], allPages, true), [item], false);
    }, []));
});
sidebarCollapsed.subscribe(function (v) {
    if (!v)
        tocCollapsed.set(true);
});
tocCollapsed.subscribe(function (v) {
    if (!v)
        sidebarCollapsed.set(true);
});
export function resolveSidebar(routeId) {
    var _a;
    if (!routeId)
        return;
    var key = Object.keys(themeOptions.sidebar || {}).find(function (key) {
        return routeId.startsWith(key);
    });
    if (key)
        resolvedSidebar.set(((_a = themeOptions.sidebar) === null || _a === void 0 ? void 0 : _a[key]) || []);
}
