// @sveltepress/theme-default/dist/vite-plugins/create-pre-core-plugins.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import process from 'node:process';
import extractorSvelte from '@unocss/extractor-svelte';
import { presetIcons, presetUno, transformerDirectives } from 'unocss';
import Unocss from 'unocss/vite';
import { SERVICE_WORKER_PATH } from '../constants.js';
import { initHighlighter } from '../markdown/highlighter.js';
var THEME_OPTIONS_MODULE = 'virtual:sveltepress/theme-default';
var DEFAULT_GRADIENT = {
    start: '#fa709a',
    end: '#fee140',
};
var DEFAULT_PRIMARY = '#fb7185';
var DEFAULT_HOVER = '#f43f5e';
function getIconSafelist(themeOptions) {
    var icons = themeOptions === null || themeOptions === void 0 ? void 0 : themeOptions.preBuildIconifyIcons;
    if (!icons)
        return [];
    var iconSafelist = [];
    var _loop_1 = function (prefix) {
        icons[prefix].forEach(function (name) {
            iconSafelist.push("i-".concat(prefix, "-").concat(name));
        });
    };
    for (var prefix in icons) {
        _loop_1(prefix);
    }
    return iconSafelist;
}
export default (function (options) { return __awaiter(void 0, void 0, void 0, function () {
    var _a, _b, gradient, _c, primary, _d, hover, iconSafelist, vitePluginsPre;
    return __generator(this, function (_e) {
        switch (_e.label) {
            case 0: return [4 /*yield*/, initHighlighter(options === null || options === void 0 ? void 0 : options.highlighter)];
            case 1:
                _e.sent();
                _a = (options === null || options === void 0 ? void 0 : options.themeColor) || {
                    gradient: DEFAULT_GRADIENT,
                    primary: DEFAULT_PRIMARY,
                    hover: DEFAULT_HOVER,
                }, _b = _a.gradient, gradient = _b === void 0 ? DEFAULT_GRADIENT : _b, _c = _a.primary, primary = _c === void 0 ? DEFAULT_PRIMARY : _c, _d = _a.hover, hover = _d === void 0 ? DEFAULT_HOVER : _d;
                iconSafelist = getIconSafelist(options);
                vitePluginsPre = [
                    Unocss({
                        extractors: [
                            extractorSvelte(),
                        ],
                        presets: [
                            presetUno(),
                            presetIcons(),
                        ],
                        transformers: [transformerDirectives()],
                        theme: {
                            colors: {
                                svp: {
                                    primary: primary,
                                    hover: hover,
                                },
                            },
                            breakpoints: {
                                sm: '950px',
                                md: '1240px',
                            },
                        },
                        shortcuts: {
                            'svp-gradient-bg': "bg-gradient-linear bg-gradient-[45deg,".concat(gradient.start, ",").concat(gradient.end, "]"),
                            'svp-gradient-text': 'svp-gradient-bg bg-clip-text text-transparent',
                            'svp-modal-bg': 'sm:hidden fixed top-0 bottom-0 right-0 left-0 bg-black dark:bg-white bg-opacity-70 dark:bg-opacity-70 z-900 opacity-0 pointer-events-none transition-opacity transition-300',
                            'svp-modal-bg-show': 'opacity-100 pointer-events-auto',
                        },
                        safelist: __spreadArray([], iconSafelist, true),
                    }),
                    {
                        name: '@sveltepress/default-theme',
                        resolveId: function (id) {
                            if (id === THEME_OPTIONS_MODULE)
                                return THEME_OPTIONS_MODULE;
                        },
                        load: function (id) {
                            if (id === THEME_OPTIONS_MODULE)
                                return "export default ".concat(JSON.stringify(options || {}));
                        },
                        config: function () {
                            return __awaiter(this, void 0, void 0, function () {
                                return __generator(this, function (_a) {
                                    return [2 /*return*/, {
                                            define: {
                                                'process.env.NODE_ENV': process.env.NODE_ENV === 'production'
                                                    ? '"production"'
                                                    : '"development"',
                                            },
                                            optimizeDeps: {
                                                exclude: ['@sveltepress/theme-default'],
                                            },
                                            server: {
                                                fs: {
                                                    allow: [SERVICE_WORKER_PATH, 'node_modules/@sveltepress/theme-default/dist/fonts', '../theme-default/dist/fonts'],
                                                },
                                            },
                                        }];
                                });
                            });
                        },
                    },
                ];
                return [2 /*return*/, vitePluginsPre];
        }
    });
}); });
